patches/minecraft/net/minecraft/world/gen/feature/jigsaw/JigsawPattern.java.patch

```diff
@@ -0,0 +1,67 @@
+--- a/net/minecraft/world/gen/feature/jigsaw/JigsawPattern.java
++++ b/net/minecraft/world/gen/feature/jigsaw/JigsawPattern.java
+@@ -21,7 +21,7 @@
+    public static final JigsawPattern field_214949_a = new JigsawPattern(new ResourceLocation("empty"), new ResourceLocation("empty"), ImmutableList.of(), JigsawPattern.PlacementBehaviour.RIGID);
+    public static final JigsawPattern field_214950_b = new JigsawPattern(new ResourceLocation("invalid"), new ResourceLocation("invalid"), ImmutableList.of(), JigsawPattern.PlacementBehaviour.RIGID);
+    private final ResourceLocation field_214951_c;
+-   private final ImmutableList<Pair<JigsawPiece, Integer>> field_214952_d;
++   private final List<Pair<JigsawPiece, Integer>> field_214952_d;
+    private final List<JigsawPiece> field_214953_e;
+    private final ResourceLocation field_214954_f;
+    private final JigsawPattern.PlacementBehaviour field_214955_g;
+@@ -29,17 +29,13 @@
+ 
+    public JigsawPattern(ResourceLocation p_i51397_1_, ResourceLocation p_i51397_2_, List<Pair<JigsawPiece, Integer>> p_i51397_3_, JigsawPattern.PlacementBehaviour p_i51397_4_) {
+       this.field_214951_c = p_i51397_1_;
+-      this.field_214952_d = ImmutableList.copyOf(p_i51397_3_);
++      this.field_214952_d = Lists.newArrayList(p_i51397_3_);
+       this.field_214953_e = Lists.newArrayList();
+ 
+-      for(Pair<JigsawPiece, Integer> pair : p_i51397_3_) {
+-         for(Integer integer = 0; integer < pair.getSecond(); integer = integer + 1) {
+-            this.field_214953_e.add(pair.getFirst().func_214845_a(p_i51397_4_));
+-         }
+-      }
+-
+       this.field_214954_f = p_i51397_2_;
+       this.field_214955_g = p_i51397_4_;
++
++      buildList();
+    }
+ 
+    public int func_214945_a(TemplateManager p_214945_1_) {
+@@ -56,6 +52,34 @@
+       return this.field_214954_f;
+    }
+ 
++   /**
++    * use {@link net.minecraftforge.event.world.jigsaw.JigsawStructurePoolInitEvent} for adding buildings
++    */
++   @Deprecated
++   public void addBuildings(List<Pair<JigsawPiece,Integer>> pieces) {
++      this.field_214952_d.addAll(pieces);
++      buildList();
++   }
++
++   @Deprecated
++   public void buildList() {
++      field_214953_e.clear();
++      for(Pair<JigsawPiece, Integer> e: field_214952_d){
++         for(Integer integer = 0; integer < e.getSecond(); integer = integer + 1) {
++            this.field_214953_e.add(e.getFirst().func_214845_a(field_214955_g));
++         }
++      }
++   }
++
++   /**
++    * use {@link net.minecraftforge.event.world.jigsaw.JigsawStructurePoolInitEvent} for removing buildings
++    */
```

Comment:
If you're telling people to use the event, then dont expose this.

Reply:
how else should this work? just comment see {@link net.minecraftforge.event.world.JigsawStructurePoolInitEvent}.
if not public the JigsawPiece List must be public?

Reply:
if the fields are public, we dont need addBuilding/buildList/removeBuilding in JigsawPattern

Reply:
Why can't you fire the event at construction time of the JigsawPatterns? Yes that'd be more events, but firing one event per Pattern shouldn't be too bad, especially as that's done only once. This way you wouldn't expose anything, you wouldn't need to iterate over stuff after the event, etc. You'd just pass the fresh created array list into the event, allow people to modify that and then apply that to the field.

(That'd be one solution which wouldn't make stuff public that shouldn't be public - there are certainly more)